oak_remote_attestation_amd/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! This crate implements remote attestation primitives from `oak_remote_attestation` based on
+//! AMD-SEV-SNP.
+//!
+//! Currently only contains some placeholder structs, which will be replaced with actual
+//! functionality as part of #1867.
+
+#![no_std]
+
+extern crate alloc;
+
+use alloc::{
+    string::{String, ToString},
+    vec::Vec,
+};
+use oak_remote_attestation::handshaker::{AttestationGenerator, AttestationVerifier};
+use serde::{Deserialize, Serialize};
+
+// TODO(#1867): Add remote attestation support and use real TEE reports.
+#[derive(Clone)]
+pub struct PlaceholderAmdAttestationGenerator;
+
+impl AttestationGenerator for PlaceholderAmdAttestationGenerator {
+    fn generate_attestation(&self, attested_data: &[u8]) -> anyhow::Result<Vec<u8>> {
+        PlaceholderAmdReport::new(attested_data)
+            .to_string()
+            .map(|s| s.as_bytes().to_vec())
+    }
+}
+
+// TODO(#1867): Add remote attestation support and use real TEE reports.
+#[derive(Clone)]
+pub struct PlaceholderAmdAttestationVerifier;
+
+impl AttestationVerifier for PlaceholderAmdAttestationVerifier {
+    fn verify_attestation(
+        &self,
+        attestation: &[u8],
+        expected_attested_data: &[u8],
+    ) -> anyhow::Result<()> {
+        let attestation_report = PlaceholderAmdReport::from_string(
+            core::str::from_utf8(attestation)
+                .map_err(|_err| anyhow::anyhow!("could not parse remote attestation report"))?,
+        )?;
+        let actual_attested_data = attestation_report.data;
+        if actual_attested_data == expected_attested_data {
+            Ok(())
+        } else {
+            Err(anyhow::anyhow!(
+                "invalid attested data; got: {:?}, expected: {:?}",
+                actual_attested_data,
+                expected_attested_data
+            ))
+        }
+    }
+}
+
+/// Placeholder implementation of TEE report for remote attestation.
+///
+/// <https://www.amd.com/system/files/TechDocs/56860.pdf#page=39>
+///
+/// TODO(#1867): Add remote attestation support and use real TEE reports.
+#[derive(Deserialize, Serialize, Debug, Default, PartialEq)]
+pub struct PlaceholderAmdReport {
+    /// Version number of this attestation report.
+    pub version: u8,
+    /// Security version number of SNP firmware.
+    pub svn: u8,
+    /// The installed version of firmware.
+    pub platform_version: u8,
+    /// Arbitrary data to put into the TEE report.
+    pub data: Vec<u8>,
+    /// TEE measurement, i.e. VM hash.
+    pub measurement: Vec<u8>,
+    /// Signature of this report.
+    pub signature: Vec<u8>,
+}
+
+// TODO(#1867): Add remote attestation support.
```

Comment:
#1867 has been closed, so should probalby not exist in TODOs (multiple places throughout the PR). A new issue (#2842) has been created to cover most of it.

Reply:
Done.